lib/rpmug.c

```diff
@@ -129,119 +143,97 @@ static int lookup_str(const char *path, long val, int vcol, int rcol,
 
 int rpmugUid(const char * thisUname, uid_t * uid)
 {
-    static char * lastUname = NULL;
-    static uid_t lastUid;
-
-    if (!thisUname) {
-	lastUname = rfree(lastUname);
-	return -1;
-    } else if (rstreq(thisUname, UID_0_USER)) {
+    if (rstreq(thisUname, UID_0_USER)) {
 	*uid = 0;
 	return 0;
     }
 
-    if (lastUname == NULL || !rstreq(thisUname, lastUname)) {
+    rpmugInit();
+
+    if (rpmug->lastUname == NULL || !rstreq(thisUname, rpmug->lastUname)) {
 	long id;
 	if (lookup_num(pwfile(), thisUname, 0, 2, &id))
 	    return -1;
-
-	free(lastUname);
-	lastUname = xstrdup(thisUname);
-	lastUid = id;
+	free(rpmug->lastUname);
+	rpmug->lastUname = xstrdup(thisUname);
+	rpmug->lastUid = id;
     }
 
-    *uid = lastUid;
+    *uid = rpmug->lastUid;
 
     return 0;
 }
 
 int rpmugGid(const char * thisGname, gid_t * gid)
 {
-    static char * lastGname = NULL;
-    static gid_t lastGid;
-
-    if (thisGname == NULL) {
-	lastGname = rfree(lastGname);
-	return -1;
-    } else if (rstreq(thisGname, GID_0_GROUP)) {
+    if (rstreq(thisGname, GID_0_GROUP)) {
 	*gid = 0;
 	return 0;
     }
 
-    if (lastGname == NULL || !rstreq(thisGname, lastGname)) {
+    rpmugInit();
+
+    if (rpmug->lastGname == NULL || !rstreq(thisGname, rpmug->lastGname)) {
 	long id;
 	if (lookup_num(grpfile(), thisGname, 0, 2, &id))
 	    return -1;
-	free(lastGname);
-	lastGname = xstrdup(thisGname);
-	lastGid = id;
+	free(rpmug->lastGname);
+	rpmug->lastGname = xstrdup(thisGname);
+	rpmug->lastGid = id;
     }
 
-    *gid = lastGid;
+    *gid = rpmug->lastGid;
 
     return 0;
 }
 
 const char * rpmugUname(uid_t uid)
 {
-    static uid_t lastUid = (uid_t) -1;
-    static char * lastUname = NULL;
-
-    if (uid == (uid_t) -1) {
-	lastUid = (uid_t) -1;
```

Comment:
I wonder if this line shouldn't be kept? `rpmugFree()` won't reset `lastUid` to `-1` and it can be used before assignment.

Reply:
Ah, this is not needed now that we have a `struct`, of course. Nevermind the above :sweat_smile: